src/Kestrel.Core/Internal/Http2/Http2Connection.cs

```diff
@@ -722,12 +788,28 @@ private Task DecodeHeadersAsync<TContext>(IHttpApplication<TContext> application
 
                 if (endHeaders)
                 {
-                    StartStream(application);
+                    lock (StateLock)
+                    {
+                        if (State != ConnectionState.Closed)
+                        {
+                            StartStream(application);
+                            HighestOpenedStreamId = _currentHeadersStream.StreamId;
```

Comment:
That's what I've been trying to say. The two lock blocks are unnecessary if we just assign first.

Reply:
Ah fair enough, reordering would work, I must have misunderstood your previous comment.

Reply:
Now I recall why I couldn't do this. We need to assign the HighestOpenedStreamId when _hpackDecoder throws stream errors. In those cases, we still need need to atomically check the state and assign the stream id. I don't see how we can simplify since in the non-error case we need an atomic check between state, stream id and starting the stream.